src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp

```diff
@@ -70,21 +75,26 @@ QVariant QgsValueRelationWidgetWrapper::value() const
       }
     }
 
-    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map )
+    QVariantList vl;
+    //store as QVariantList because it's json
+    for ( const QString &s : qgis::as_const( selection ) )
     {
-      QVariantList vl;
-      //store as QVariantList because it's json
-      for ( const QString &s : qgis::as_const( selection ) )
+      // Convert to proper type
+      const auto type { fkType() };
```

Comment:
Might be a personal opinion, I prefer explicit over implicit because that improves code readability (i.e. I only use auto when it's very clear what it's about and the real type is crappy to write like e.g. iterators).

Reply:
Well, the good thing with `auto` is that if for some reason we change the return type in the future it *may* not be necessary to change other parts of the code.
I see your point, but since QtCreator show a nice tooltip with the actual type I think it's not big issue.

Reply:
The point is, github doesn't ;)

Reply:
Also, I may be alone in this, but to me the ability to change return type without compilation errors is a huge step backwards. There's no telling what subtle bugs may arise as a result of doing this.

Reply:
Ok guys you have almost convinced me, I will use less auto.
 Pity because l was starting to like C++ ;)

Reply:
Just to further explain my concerns:

Let's say I have the c++ method

    QgsGeometry getValue() const;

It's well tested under python, with lots of unit tests covering all possible return values. In future someone comes along and decides that this method needs to be more flexible (and e.g. sometimes return QgsRectangles), so they change the signature:

    QVariant getValue() const;

All the existing python tests are fine, because sip auto-converts the QVariant return values to QgsGeometry. So it looks good, nothing has broken.

But buried somewhere in c++ is this usage:

   auto val = foo.getValue();
   if ( !val.isNull() )
   {
      ....
    }

Previously, this was testing QgsGeometry::isNull(), and worked as intended. By changing the return type to QVariant, this block is now testing QVariant::isNull(), which has a very different meaning! Bam, silently, very difficult to diagnose regression, despite being soaked by Python unit tests.

That's why auto is a step back -- if we used QgsGeometry = foo.getValue(), compilation would break and we'd see that that block needs fixing manually.

Reply:
Thanks for the explanation.
I think we need a 'soft' policy.
Would everyone agrees on this answer?
https://stackoverflow.com/a/6435036?stw=2

Reply:
+1 for the soft policy.

> I think that one should use the auto keyword whenever it's hard to say how to write the type at first sight, but the type of the right hand side of an expression is obvious.

Just one minor correction, where I think there is a mismatch between examples and the above specification.

```cpp
std::shared_ptr<Foo> foo = std::make_shared<Foo>(); // while obvious, it's also easy to write at first sight
```